File: threads/src/main/java/com/sequenia/threads/activities/GalleryActivity.java
package com.sequenia.threads.activities;

import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.database.Cursor;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.provider.MediaStore;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;

import com.sequenia.threads.BucketGalleryDecorator;
import com.sequenia.threads.GalleryDecorator;
import com.sequenia.threads.R;
import com.sequenia.threads.adapters.GalleryAdaper;
import com.sequenia.threads.adapters.PhotoBucketsGalleryAdapter;
import com.sequenia.threads.model.MediaPhoto;
import com.sequenia.threads.model.PhotoBucketItem;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by yuri on 06.07.2016.
 */
public class GalleryActivity extends AppCompatActivity
        implements
        PhotoBucketsGalleryAdapter.OnItemClick
        , GalleryAdaper.OnGalleryItemClick {
    private RecyclerView mRecyclerView;
    private static final String TAG = "GalleryActivity ";
    private ArrayList<ArrayList<MediaPhoto>> lists;
    List<PhotoBucketItem> bucketItems = new ArrayList<>();
    private boolean isInBuckets = false;
    private BucketGalleryDecorator mBucketGalleryDecorator = new BucketGalleryDecorator(4);
    private GalleryDecorator mGalleryDecorator = new GalleryDecorator(4);
    private EditText mSearchEdiText;
    private List<MediaPhoto> chosentItems;
    private Button mSendButton;
    public static final String PHOTOS_REQUEST_CODE_TAG = "PHOTOS_REQUEST_CODE_TAG";
    public static final String PHOTOS_TAG = "PHOTOS_TAG";


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_gallery);
        initViews();
    }

    private void initViews() {
        final Context ctx = this;
        mRecyclerView = (RecyclerView) findViewById(R.id.recycler);

        findViewById(R.id.search_photo).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                setStateSearchingPhoto();
            }
        });
        Toolbar t = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(t);
        t.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onBackPressed();
            }
        });
        //  t.showOverflowMenu();
        Drawable overflowDrawable = t.getOverflowIcon();
        try {
            overflowDrawable.setColorFilter(new PorterDuffColorFilter(getResources().getColor(android.R.color.white), PorterDuff.Mode.SRC_ATOP));
        } catch (Resources.NotFoundException e) {
            e.printStackTrace();
        }
        mRecyclerView = (RecyclerView) findViewById(R.id.recycler);
        mSendButton = (Button) findViewById(R.id.send);
        String[] projection = new String[]{MediaStore.Images.Media.BUCKET_DISPLAY_NAME, MediaStore.Images.Media.DATA};
        Cursor c = getContentResolver().query(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, projection, null, null, MediaStore.Images.Media.BUCKET_DISPLAY_NAME + " desc");
        int BUCKET_DISPLAY_NAME = c.getColumnIndex(MediaStore.Images.Media.BUCKET_DISPLAY_NAME);
        int DATA = c.getColumnIndex(MediaStore.Images.Media.DATA);
        if (c.getCount() == 0) return;
        ArrayList<MediaPhoto> allItems = new ArrayList<>();
        for (c.moveToFirst(); !c.isAfterLast(); c.moveToNext()) {
            allItems.add(new MediaPhoto(c.getString(DATA), c.getString(BUCKET_DISPLAY_NAME)));
        }
        lists = new ArrayList<>();
        ArrayList<MediaPhoto> list = new ArrayList<>();
        list.add(allItems.get(0));
        lists.add(list);
        for (int i = 1; i < allItems.size(); i++) {
            if (allItems.get(i - 1).getBucketName().equalsIgnoreCase(allItems.get(i).getBucketName())) {
                for (int j = 0; j < lists.size(); j++) {
                    if (lists.get(j).get(0).getBucketName().equalsIgnoreCase(allItems.get(i).getBucketName())) {
                        lists.get(j).add(allItems.get(i));
                        break;
                    }
                }
            } else {
                list = new ArrayList<>();
                list.add(allItems.get(i));
                lists.add(list);
            }
        }
        for (ArrayList<MediaPhoto> itemList : lists) {
            bucketItems.add(new PhotoBucketItem(itemList.get(0).getBucketName(), String.valueOf(itemList.size()), itemList.get(0).getImagePath()));
        }
        mBucketGalleryDecorator = new BucketGalleryDecorator(4);
        mRecyclerView.setLayoutManager(new GridLayoutManager(this, 2));
        mRecyclerView.setAdapter(new PhotoBucketsGalleryAdapter(bucketItems, this));
        mRecyclerView.addItemDecoration(mBucketGalleryDecorator);
        isInBuckets = true;
        mSearchEdiText = (EditText) findViewById(R.id.search_edit_text);
        findViewById(R.id.cancel).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onBackPressed();
            }
        });
        mSendButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (getIntent().getIntExtra(PHOTOS_REQUEST_CODE_TAG, -1) == -1) {
                    finish();
                } else {
                    ArrayList<String> list1 = new ArrayList<String>();
                    for (MediaPhoto mp : chosentItems) {
                        list1.add(mp.getImagePath());
                    }
                    Intent i = new Intent();
                    i.putStringArrayListExtra(PHOTOS_TAG, list1);
                    setResult(RESULT_OK, i);
                    finish();
                }
            }
        });
    }

    public static Intent getStartIntent(Context ctx, int requestCode) {
        Intent i = new Intent(ctx, GalleryActivity.class);
        i.putExtra(PHOTOS_REQUEST_CODE_TAG, requestCode);
        return i;
    }

    @Override
    public void onPhotoBucketClick(PhotoBucketItem item) {
        setStateGallery(item.getBucketName(), item);
    }

    private void setStateGallery(String title, PhotoBucketItem item) {
        chosentItems = null;
        isInBuckets = false;
        checkSendButtonState();
        ((Toolbar) findViewById(R.id.toolbar)).setTitle(title);
        findViewById(R.id.search_label_layout).setVisibility(View.GONE);
        findViewById(R.id.bottom_buttons).setVisibility(View.VISIBLE);
        findViewById(R.id.search_layout).setVisibility(View.GONE);
        findViewById(R.id.nothing_found_label).setVisibility(View.GONE);
        mRecyclerView.removeItemDecoration(mBucketGalleryDecorator);
        mRecyclerView.setLayoutManager(new GridLayoutManager(this, 3));
        List<MediaPhoto> photos = null;
        for (List<MediaPhoto> list : lists) {
            if (list.get(0).getImagePath().equals(item.getImagePath())) {
                photos = list;
                break;
            }
        }
        for (MediaPhoto photo : photos) {
            photo.setChecked(false);
        }
        mRecyclerView.setAdapter(new GalleryAdaper(photos, this));
        mRecyclerView.addItemDecoration(mGalleryDecorator);

    }

    private void setStatePhotoBuckets() {
        isInBuckets = true;
        chosentItems = null;
        ((Toolbar) findViewById(R.id.toolbar)).setTitle(getResources().getString(R.string.photos));
        findViewById(R.id.search_label_layout).setVisibility(View.VISIBLE);
        findViewById(R.id.bottom_buttons).setVisibility(View.GONE);
        findViewById(R.id.search_layout).setVisibility(View.GONE);
        findViewById(R.id.nothing_found_label).setVisibility(View.GONE);
        ((Toolbar) findViewById(R.id.toolbar)).showOverflowMenu();
        mRecyclerView.removeItemDecoration(mGalleryDecorator);
        mRecyclerView.setLayoutManager(new GridLayoutManager(this, 2));
        mRecyclerView.setAdapter(new PhotoBucketsGalleryAdapter(bucketItems, this));
        mRecyclerView.addItemDecoration(mBucketGalleryDecorator);
    }

    private void setStateSearchingPhoto() {
        isInBuckets = false;
        chosentItems = null;
        checkSendButtonState();
        findViewById(R.id.search_label_layout).setVisibility(View.GONE);
        findViewById(R.id.search_layout).setVisibility(View.VISIBLE);
        mSearchEdiText.requestFocus();
        ImageButton clearButton = (ImageButton) findViewById(R.id.clear_search_button);
        final TextView nothingFoundLabel = (TextView) findViewById(R.id.nothing_found_label);
        nothingFoundLabel.setVisibility(View.VISIBLE);
        mRecyclerView.removeItemDecoration(mBucketGalleryDecorator);
        mRecyclerView.addItemDecoration(mGalleryDecorator);
        mRecyclerView.setLayoutManager(new GridLayoutManager(this, 3));
        mRecyclerView.setAdapter(null);
        final View bottomButton = findViewById(R.id.bottom_buttons);
        clearButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mSearchEdiText.setText("");
            }
        });
        final GalleryAdaper.OnGalleryItemClick listener = this;
        bottomButton.setVisibility(View.GONE);
        ((Toolbar) findViewById(R.id.toolbar)).hideOverflowMenu();
        mSearchEdiText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                clearCheckedStateOfItems();
                chosentItems = null;
                checkSendButtonState();
                if (s.length() == 0) {
                    nothingFoundLabel.setVisibility(View.VISIBLE);
                    mRecyclerView.setAdapter(null);
                    bottomButton.setVisibility(View.GONE);
                } else {
                    nothingFoundLabel.setVisibility(View.GONE);
                    List<MediaPhoto> list = new ArrayList<>();
                    for (List<MediaPhoto> photoes : lists) {
                        for (MediaPhoto mp : photoes) {
                            if (mp.getImagePath().contains(s.toString())) {
                                list.add(mp);
                            }
                        }
                    }
                    mRecyclerView.setAdapter(new GalleryAdaper(list, listener));
                    if (list.size() == 0) {
                        bottomButton.setVisibility(View.GONE);
                    } else {
                        bottomButton.setVisibility(View.VISIBLE);
                    }
                }
            }
        });
    }

    @Override
    public void onGalleryItemsChosen(List<MediaPhoto> chosenItems) {
        this.chosentItems = chosenItems;
        checkSendButtonState();
    }

    @Override
    public void onBackPressed() {
        if (!isInBuckets) {
            mSearchEdiText.setText("");
            clearCheckedStateOfItems();
            findViewById(R.id.nothing_found_label).setVisibility(View.GONE);
            setStatePhotoBuckets();
        } else {
            super.onBackPressed();
        }
    }

    private void clearCheckedStateOfItems() {
        for (List<MediaPhoto> list : lists) {
            for (MediaPhoto mp : list) {
                mp.setChecked(false);
            }
        }
    }

    private void checkSendButtonState() {
        if (null != chosentItems && chosentItems.size() > 0) {
            mSendButton.setEnabled(true);
            mSendButton.setTextColor(getResources().getColor(android.R.color.white));
        } else {
            mSendButton.setEnabled(false);
            mSendButton.setTextColor(getResources().getColor(R.color.disabled_text_color));
        }
    }
}

File: src/main/java/im/threads/ui/activities/GalleryActivity.kt
package im.threads.ui.activities

import android.content.ContentUris
import android.content.Context
import android.content.Intent
import android.graphics.drawable.ColorDrawable
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.text.Editable
import android.text.TextWatcher
import android.view.Gravity
import android.view.MenuItem
import android.view.inputmethod.EditorInfo
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.GridLayoutManager
import im.threads.R
import im.threads.business.extensions.withMainContext
import im.threads.business.logger.LoggerEdna
import im.threads.business.models.MediaPhoto
import im.threads.business.models.PhotoBucketItem
import im.threads.business.utils.MediaHelper
import im.threads.databinding.EccActivityGalleryBinding
import im.threads.ui.adapters.GalleryAdapter
import im.threads.ui.adapters.GalleryAdapter.OnGalleryItemClick
import im.threads.ui.adapters.PhotoBucketsGalleryAdapter
import im.threads.ui.adapters.PhotoBucketsGalleryAdapter.OnItemClick
import im.threads.ui.config.Config.Companion.getInstance
import im.threads.ui.utils.BucketsGalleryDecorator
import im.threads.ui.utils.ColorsHelper
import im.threads.ui.utils.ColorsHelper.setDrawableColor
import im.threads.ui.utils.GalleryDecorator
import im.threads.ui.utils.gone
import im.threads.ui.utils.visible
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.BehaviorSubject
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import java.util.Locale

class GalleryActivity : BaseActivity(), OnItemClick, OnGalleryItemClick {

    private val screenState = BehaviorSubject.createDefault(ScreenState.BUCKET_LIST)
    private val dataIsEmpty = BehaviorSubject.createDefault(false)

    private var photosMap: MutableMap<String, MutableList<MediaPhoto>> = HashMap()
    private val bucketItems: MutableList<PhotoBucketItem> = ArrayList()
    private val chosenItems: MutableList<MediaPhoto> = ArrayList()
    private val bucketsGalleryDecorator = BucketsGalleryDecorator(4)
    private val galleryDecorator = GalleryDecorator(4)
    private val coroutineScope = CoroutineScope(Dispatchers.IO)
    private val compositeDisposable = CompositeDisposable()

    private lateinit var binding: EccActivityGalleryBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = EccActivityGalleryBinding.inflate(layoutInflater)
        setContentView(binding.root)

        initViews()
        subscribeToScreenState()
        subscribeToDataIsEmpty()
        initStatusBar()
        coroutineScope.launch {
            initData()
            withMainContext {
                showBucketListState()
                setButtonsClickListeners()
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        compositeDisposable.clear()
    }

    private fun initStatusBar() {
        val style = getInstance().chatStyle
        val statusBarColor = ContextCompat.getColor(this, style.chatStatusBarColorResId)
        val toolBarColor = ContextCompat.getColor(this, style.chatToolbarColorResId)
        val isStatusBarLight =
            resources.getBoolean(getInstance().chatStyle.windowLightStatusBarResId)
        super.setStatusBarColor(isStatusBarLight, statusBarColor)
        val backButtonDrawable =
            AppCompatResources.getDrawable(this, R.drawable.ecc_ic_arrow_back_white_24dp)
        setDrawableColor(this, backButtonDrawable, style.chatToolbarTextColorResId)
        supportActionBar?.apply {
            setBackgroundDrawable(ColorDrawable(toolBarColor))
            setHomeAsUpIndicator(style.chatToolbarBackIconResId)
            setHomeAsUpIndicator(backButtonDrawable)
        }
        val textColor = ContextCompat.getColor(this, style.chatToolbarTextColorResId)
        val hintTextColor = ContextCompat.getColor(this, style.chatToolbarHintTextColor)
        binding.toolbar.title
        binding.searchEditText.setTextColor(textColor)
        binding.searchEditText.setHintTextColor(hintTextColor)
        binding.clearSearchButton.setImageResource(R.drawable.ecc_ic_clear_gray_30dp)
        binding.title.setTextColor(textColor)
        ColorsHelper.setTint(this, binding.clearSearchButton, style.chatToolbarTextColorResId)
        ColorsHelper.setTint(this, binding.backButton, style.chatToolbarTextColorResId)

        initToolbarTextPosition()
    }

    private fun initToolbarTextPosition() {
        val isToolbarTextCentered = getInstance().chatStyle.isToolbarTextCentered
        val gravity = if (isToolbarTextCentered) Gravity.CENTER else Gravity.CENTER_VERTICAL
        binding.title.gravity = gravity
    }

    private fun setButtonsClickListeners() = with(binding) {
        backButton.setOnClickListener { onBackPressed() }
        clearSearchButton.setOnClickListener { clearSearch() }
        searchPhoto.setOnClickListener { showSearch() }
        cancel.setOnClickListener { onBackPressed() }
        send.setOnClickListener { send() }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return if (item.itemId == android.R.id.home) {
            onBackPressed()
            true
        } else {
            super.onOptionsItemSelected(item)
        }
    }

    override fun onBackPressed() {
        if (ScreenState.BUCKET_LIST != screenState.value) {
            binding.searchEditText.setText("")
            showBucketListState()
        } else {
            super.onBackPressed()
        }
    }

    override fun onPhotoBucketClick(item: PhotoBucketItem) {
        showPhotoListState(item.bucketName, item)
    }

    override fun onGalleryItemsChosen(chosenItems: List<MediaPhoto>) {
        this.chosenItems.clear()
        this.chosenItems.addAll(chosenItems)
        syncSendButtonState()
    }

    fun clearSearch() {
        binding.searchEditText.setText("")
    }

    fun showSearch() {
        showSearchState()
        search("")
    }

    fun send() {
        val list = ArrayList<Uri>()
        chosenItems.forEach {
            list.add(it.imageUri)
        }
        val intent = Intent()
        intent.putParcelableArrayListExtra(PHOTOS_TAG, list)
        setResult(RESULT_OK, intent)
        finish()
    }

    private fun initViews() {
        setSupportActionBar(binding.toolbar)
        binding.searchEditText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable) {
                if (s.isNotEmpty()) {
                    search(s.toString())
                } else {
                    search("")
                }
            }
        })
        binding.searchEditText.setOnEditorActionListener { v, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_SEARCH) {
                search(v.text.toString())
                return@setOnEditorActionListener true
            } else {
                return@setOnEditorActionListener false
            }
        }
    }

    private fun initData() {
        MediaHelper.getAllPhotos(this).use { cursor ->
            cursor?.let {
                if (it.count > 0) {
                    val idKey = it.getColumnIndex(MediaStore.Images.Media._ID)
                    val displayNameKey = it.getColumnIndex(MediaStore.Images.Media.DISPLAY_NAME)
                    val bucketDisplayNameKey =
                        it.getColumnIndex(MediaStore.Images.Media.BUCKET_DISPLAY_NAME)
                    photosMap = HashMap()
                    it.moveToFirst()
                    while (!it.isAfterLast) {
                        val imageUri = ContentUris.withAppendedId(
                            MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                            it.getLong(idKey)
                        )
                        val bucketName = it.getString(bucketDisplayNameKey)
                        val displayName = it.getString(displayNameKey)
                        if (!bucketName.isNullOrEmpty()) {
                            var mediaPhotos: MutableList<MediaPhoto>?
                            if (photosMap.containsKey(bucketName)) {
                                mediaPhotos = photosMap[bucketName]
                                if (mediaPhotos == null) {
                                    mediaPhotos = ArrayList()
                                }
                            } else {
                                mediaPhotos = ArrayList()
                            }
                            if (!displayName.isNullOrEmpty()) {
                                mediaPhotos.add(
                                    MediaPhoto(
                                        imageUri,
                                        displayName,
                                        bucketName
                                    )
                                )
                            }
                            if (mediaPhotos.size > 0) {
                                photosMap[bucketName] = mediaPhotos
                            }
                        }
                        it.moveToNext()
                    }
                }
            }

            for ((key, value) in photosMap) {
                if (value.size > 0) {
                    bucketItems.add(
                        PhotoBucketItem(
                            key,
                            value.size.toString(),
                            value[0].imageUri
                        )
                    )
                }
            }
        }
    }

    private fun showBucketListState() {
        screenState.onNext(ScreenState.BUCKET_LIST)
        chosenItems.clear()
        setTitle(resources.getString(R.string.ecc_photos))
        binding.recycler.removeItemDecoration(galleryDecorator)
        binding.recycler.addItemDecoration(bucketsGalleryDecorator)
        binding.recycler.layoutManager = GridLayoutManager(this, 2)
        binding.recycler.adapter = PhotoBucketsGalleryAdapter(bucketItems, this)
        dataIsEmpty.onNext(bucketItems.isEmpty())
    }

    private fun showPhotoListState(title: String, item: PhotoBucketItem) {
        screenState.onNext(ScreenState.PHOTO_LIST)
        chosenItems.clear()
        syncSendButtonState()
        setTitle(title)
        binding.recycler.removeItemDecoration(bucketsGalleryDecorator)
        binding.recycler.addItemDecoration(galleryDecorator)
        binding.recycler.layoutManager = GridLayoutManager(this, 3)
        var photos: List<MediaPhoto>? = null
        for (list in photosMap.values) {
            if (list[0].imageUri == item.imagePath) {
                photos = list
                break
            }
        }
        photos?.forEach {
            it.isChecked = false
        }
        binding.recycler.adapter = GalleryAdapter(photos, this)
        dataIsEmpty.onNext(photos.isNullOrEmpty())
    }

    private fun showSearchState() {
        screenState.onNext(ScreenState.SEARCH)
        chosenItems.clear()
        syncSendButtonState()
        binding.searchEditText.requestFocus()
        binding.recycler.removeItemDecoration(bucketsGalleryDecorator)
        binding.recycler.addItemDecoration(galleryDecorator)
        binding.recycler.layoutManager = GridLayoutManager(this, 3)
        binding.recycler.adapter = null
        dataIsEmpty.onNext(true)
    }

    private fun search(searchString: String) {
        clearCheckedStateOfItems()
        chosenItems.clear()
        syncSendButtonState()
        val list: MutableList<MediaPhoto> = ArrayList()
        photosMap.values.forEach {
            it.forEach { photo ->
                if (photo.imageUri.toString().lowercase(Locale.getDefault()).contains(
                        searchString.lowercase(Locale.getDefault())
                    ) ||
                    photo.displayName.contains(searchString.lowercase(Locale.getDefault()))
                ) {
                    list.add(photo)
                }
            }
        }
        binding.recycler.adapter = GalleryAdapter(list, this)
        dataIsEmpty.onNext(list.isEmpty())
    }

    private fun clearCheckedStateOfItems() {
        photosMap.values.forEach {
            it.forEach { photo ->
                photo.isChecked = false
            }
        }
    }

    private fun syncSendButtonState() {
        if (chosenItems.size > 0) {
            binding.send.isEnabled = true
            binding.send.setTextColor(ContextCompat.getColor(this, android.R.color.white))
        } else {
            binding.send.isEnabled = false
            binding.send.setTextColor(
                ContextCompat.getColor(
                    this,
                    R.color.ecc_disabled_text_color
                )
            )
        }
    }

    private fun subscribeToScreenState() {
        compositeDisposable.add(
            screenState
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    if (it == ScreenState.SEARCH) {
                        binding.searchLayout.visible()
                    } else {
                        binding.searchLayout.gone()
                    }

                    if (it == ScreenState.BUCKET_LIST) {
                        binding.searchLabelLayout.visible()
                    } else {
                        binding.searchLabelLayout.gone()
                    }

                    checkBottomButtons()
                }, {
                    LoggerEdna.error(it)
                })
        )
    }

    private fun subscribeToDataIsEmpty() {
        compositeDisposable.add(
            dataIsEmpty
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ isDataEmpty ->
                    if (isDataEmpty) {
                        binding.nothingFoundLabel.visible()
                    } else {
                        binding.nothingFoundLabel.gone()
                    }
                    checkBottomButtons()
                }, {
                    LoggerEdna.error(it)
                })
        )
    }

    private fun checkBottomButtons() {
        if (dataIsEmpty.value != true && screenState.value != ScreenState.BUCKET_LIST) {
            binding.bottomButtons.visible()
        } else {
            binding.bottomButtons.gone()
        }
    }

    enum class ScreenState {
        BUCKET_LIST, PHOTO_LIST, SEARCH
    }

    companion object {
        const val PHOTOS_TAG = "PHOTOS_TAG"
        private const val PHOTOS_REQUEST_CODE_TAG = "PHOTOS_REQUEST_CODE_TAG"

        @JvmStatic
        fun getStartIntent(ctx: Context?, requestCode: Int): Intent {
            val intent = Intent(ctx, GalleryActivity::class.java)
            intent.putExtra(PHOTOS_REQUEST_CODE_TAG, requestCode)
            return intent
        }
    }
}
